### PR TITLE
Add decoding of contract log events

### PR DESCRIPTION
**Pull request by DOBEN**
## Purpose

Add decoding of contract log events in interrupt chain events

## Changes

Add decoding of contract log events in interrupt chain events

### CODE REVIEW

File: backend-rust/src/graphql_api.rs [hunk 1]
```diff
@@ -5061,15 +5062,70 @@ pub struct ChainUpdatePayload {
...
+        let mut connection = connection::Connection::new(true, true);
+
+        for (index, log) in self.contract_logs_raw.iter().enumerate() {
+            let decoded_log =
+                decode_value_with_schema(opt_event_schema.as_ref(), "event", log, "contract_log");
+
+            connection.edges.push(connection::Edge::new(index.to_string(), decoded_log));
+        }
+
+        // Nice-to-have: pagination info but not used at front-end currently.
```
[the remark this answers is not in the document]
Comment (DOBEN) [reply backend-rust/src/graphql_api.rs:5125]: Currently, the front-end does not paginate the `contract_logs` (usually there are only a few) and this backend just uses some `default` pagination setting when generating the `connection` for the `contract_logs` as such.

```
let mut connection = connection::Connection::new(true, true);
```

File: backend-rust/src/graphql_api.rs [hunk 2]
```diff
@@ -5061,15 +5062,70 @@ pub struct ChainUpdatePayload {
...
+        .fetch_optional(pool)
+        .await?
+        .ok_or(ApiError::NotFound)?;
+
+        let opt_event_schema = row
+            .display_schema
+            .as_ref()
+            .and_then(|schema| VersionedModuleSchema::new(schema, &None).ok())
+            .and_then(|versioned_schema| {
+                versioned_schema.get_event_schema(&row.contract_name).ok()
```
Comment (lassemand) [inline backend-rust/src/graphql_api.rs:5113]: Same for this and should probably remove curly brackets for consistency

File: backend-rust/src/graphql_api.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (DOBEN) [reply backend-rust/src/graphql_api.rs:5125]: I changed the inline comment to `TODO`. I feel it is important to know why `pagination` is not handled here at the moment when reading the code.

File: backend-rust/src/graphql_api.rs [hunk 3]
```diff
@@ -5061,15 +5062,70 @@ pub struct ChainUpdatePayload {
...
+            });
+
+        let mut connection = connection::Connection::new(true, true);
+
+        for (index, log) in self.contract_logs_raw.iter().enumerate() {
+            let decoded_log =
+                decode_value_with_schema(opt_event_schema.as_ref(), "event", log, "contract_log");
```
Comment (lassemand) [inline backend-rust/src/graphql_api.rs:5120]: I assume "event" should be the display_name?

At the very least it seems to be unused in the method besides providing error messages. And for the error message it seem to say that no schema with such name exist?

File: backend-rust/src/graphql_api.rs (hunk 1 above)
Comment (lassemand) [inline backend-rust/src/graphql_api.rs:5125]:
```suggestion
        // TODO: pagination info but not used at front-end currently.
```

Comment (DOBEN) [reply backend-rust/src/graphql_api.rs:5125]: I use the comment. For prioritization it is important that it is currently not used and as such not so important to implement vs some other TODOS.
```
// TODO: pagination info but not used at front-end currently.
```

File: backend-rust/src/graphql_api.rs (hunk 2 above)
Comment (lassemand) [reply backend-rust/src/graphql_api.rs:5113]: Based on conversation elsewhere this should not be logging error, but curly brackets can still be removed for consistency

File: backend-rust/src/graphql_api.rs [hunk 4]
```diff
@@ -5105,24 +5152,33 @@ impl ContractInterrupted {
...
-            .and_then(|schema| VersionedModuleSchema::new(schema, &None).ok())
-            .and_then(|versioned_schema| {
-                versioned_schema.get_event_schema(&row.contract_name).ok()
-            });
+        // Get the event schema if it exists.
+        let opt_event_schema = if let Some(event_schema) = row.display_schema.as_ref() {
+            let versioned_schema =
+                VersionedModuleSchema::new(event_schema, &None).map_err(|_| {
+                    ApiError::InternalError(
+                        "Database bytes should be a valid VersionedModuleSchema".to_string(),
```
Comment (lassemand) [inline backend-rust/src/graphql_api.rs:5160]: We should definitely also include this error to the logs.

**Review by lassemand — APPROVED** [review]

File: backend-rust/src/graphql_api.rs (hunk 4 above)
Comment (DOBEN) [reply backend-rust/src/graphql_api.rs:5162]: Yes, I agree. We have an issue for building a middleware to trace all `internalError` in our service (not implemented yet). When addressing the below issues we likely will decide to log/trace more information internally (e.g. exact database connection issues) while reporting only the parts of the internal errors to the user that don't leak any problematic information.
https://github.com/Concordium/concordium-scan/issues/270

File: backend-rust/src/graphql_api.rs (hunk 3 above)
Comment (DOBEN) [reply backend-rust/src/graphql_api.rs]: Revised that part and decided to implement some prober types `SmartContractSchemaNames` for the names of schemas.

**Pull request merged**